src/Components/Web/src/HtmlRendering/StaticHtmlRenderer.HtmlWriting.cs

```diff
@@ -351,12 +351,29 @@ void EmitFormActionIfNotExplicit(TextWriter output, bool isForm, bool hasExplici
                 output.Write("action");
                 output.Write('=');
                 output.Write('\"');
-                _htmlEncoder.Encode(output, _navigationManager.Uri);
+                _htmlEncoder.Encode(output, GetRootRelativeUrlForFormAction(_navigationManager));
                 output.Write('\"');
             }
         }
     }
 
+    private static string GetRootRelativeUrlForFormAction(NavigationManager navigationManager)
+    {
+        // We want a root-relative URL because:
+        // - if we used a base-relative one, then if currentUrl==baseHref, that would result
+        //   in an empty string, but forms have special handling for action="" (it means "submit
+        //   to the current URL, but that would be wrong if there's an uncommitted navigation in
+        //   flight, e.g., after the user clicking 'back' - it would go to whatever's now in the
+        //   address bar, ignoring where the form was rendered)
+        // - if we used an absolute URL, then it creates a significant extra pit of failure for
+        //   apps hosted behind a reverse proxy (e.g., container apps), because the server's view
+        //   of the absolute URL isn't usable outside the container
+        //   - of course, sites hosted behind URL rewriting that modifies the path will still be
+        //     wrong, but developers won't do that often as it makes things like <a href> really
+        //     difficult to get right. In that case, developers must emit an action attribute manually.
+        return new Uri(navigationManager.Uri, UriKind.Absolute).PathAndQuery;
```

Comment:
Curious why you can't use the existing `Uri` instance in the private field on `NavigationManager` instead of allocating a new one?

Reply:
The `_uri` on NavigationManager is a `string` so we wouldn't gain anything. Even if it was a `System.Uri`, it's in a different assembly and we don't have IVT to it. We would have to add IVT or a new public API.

Having one extra allocation per actionless-form is not likely to be problematic, and especially at this stage in the release we need to minimize code changes. We can optimize this more in the future if needed.